app/Services/NetworkDataService.php

```diff
@@ -56,11 +56,62 @@ private function processClients(array $clients): void
      */
     private function handleTimeouts(): void
     {
-        NetworkAircraft::all()->each(function (NetworkAircraft $aircraft) {
-            if ($aircraft->updated_at < Carbon::now()->subMinutes(10)) {
-                $aircraft->delete();
-                event(new NetworkAircraftDisconnectedEvent($aircraft));
+        NetworkAircraft::all()->each(
+            function (NetworkAircraft $aircraft) {
+                if ($aircraft->updated_at < Carbon::now()->subMinutes(10)) {
+                    $aircraft->delete();
+                    event(new NetworkAircraftDisconnectedEvent($aircraft));
+                }
+            }
+        );
+    }
+
+    public static function createOrUpdateNetworkAircraft(
+        string $callsign,
+        array $details = []
+    ): NetworkAircraft {
+        try {
+            $aircraft = NetworkAircraft::updateOrCreate(
+                ['callsign' => $callsign],
+                array_merge(
+                    ['callsign' => $callsign],
+                    $details
+                )
+            );
+            $aircraft->touch();
+        } catch (QueryException $queryException) {
+            if ($queryException->errorInfo[1] !== 1062) {
+                throw $queryException;
+            }
+            $aircraft = NetworkAircraft::find($callsign);
+        }
+
+        return $aircraft;
+    }
+
+    public static function firstOrCreateNetworkAircraft(
+        string $callsign,
+        array $details = []
+    ): NetworkAircraft {
+        try {
+            $aircraft = NetworkAircraft::firstOrCreate(
+                ['callsign' => $callsign],
+                array_merge(
+                    ['callsign' => $callsign],
+                    $details
+                )
+            );
+
+            if ($aircraft->wasRecentlyCreated) {
```

Comment:
Is this meant to be a negative instead? I'm thinking that if it was recently created then the timestamps should be fine?

Reply:
Actually this bit isn't required, as I set the `updated_at` column whenever the model is created :)